src/filters/KalmanFilter.h

```diff
@@ -0,0 +1,278 @@
+	 * @param measurementStdDevs The standard deviations of the measurement. This implies
+	 * inaccuracy in the sensor.
+	 * @param dt The time in seconds between each update.
+	 * @return A Kalman Filter built for the specified system.
+	 */
+	static KalmanFilter
+	createCont(const Eigen::Matrix<double, numStates, numStates> &systemMat,
+			   const Eigen::Matrix<double, numStates, numInputs> &inputMat,
+			   const Eigen::Matrix<double, numStates, numStates> &outputMat,
+			   const Eigen::Matrix<double, numStates, 1> &stateStdDevs,
+			   const Eigen::Matrix<double, numStates, 1> &measurementStdDevs, double dt)
+	{
+		matrix stateCovarianceCont = createCovarianceMatrix(stateStdDevs);
+		matrix measurementCovarianceCont = createCovarianceMatrix(measurementStdDevs);
+
+		matrix discA = systemMat;
+		matrix discB = inputMat;
+		continuousToDiscrete(discA, discB, dt);
+
+		matrix stateCovariance = discretizeQ(systemMat, stateCovarianceCont, dt);
+		matrix measurementCovariance = discretizeR(measurementCovarianceCont, dt);
+
+		// solve DARE for asymptotic state error covariance matrix
+		matrix P = DARE(discA.transpose(), outputMat.transpose(), stateCovariance,
+						measurementCovariance);
+
+		matrix S = outputMat * P * outputMat.transpose() + measurementCovariance;
+		// This is the Kalman gain matrix, used to weight the GPS data against the model data
+		matrix gainMatrix =
+			S.transpose().colPivHouseholderQr().solve((outputMat * P.transpose()).transpose());
+
+		return KalmanFilter(discA, discB, outputMat, stateCovariance, measurementCovariance,
+							gainMatrix);
+	}
+
+	/**
+	 * Create a new Kalman Filter from a discrete-time state space model.
+	 * The only matrices that change between this and createCont() are the system matrix and
+	 * input matrix. All other parameters would be the same.
+	 *
+	 * @param systemMat The discrete-time system matrix.
+	 * @param inputMat  The discrete-time input matrix.
+	 * @param outputMat The output matrix.
+	 * @param stateStdDevs The standard deviations for the system model. This implies
+	 * inaccuracy in the state space model.
+	 * @param measurementStdDevs The standard deviations of the measurement. This implies
+	 * inaccuracy in the sensor.
+	 * @param dt The time in seconds between each update.
+	 * @return A Kalman Filter built for the specified system.
+	 */
+	static KalmanFilter
+	createDisc(const Eigen::Matrix<double, numStates, numStates> &systemMat,
+			   const Eigen::Matrix<double, numStates, numInputs> &inputMat,
+			   const Eigen::Matrix<double, numStates, numStates> &outputMat,
+			   const Eigen::Matrix<double, numStates, 1> &stateStdDevs,
+			   const Eigen::Matrix<double, numStates, 1> &measurementStdDevs, double dt)
+	{
+		matrix stateCovarianceCont = createCovarianceMatrix(stateStdDevs);
+		matrix measurementCovarianceCont = createCovarianceMatrix(measurementStdDevs);
+
+		matrix contA = systemMat;
+		matrix contB = inputMat;
+		discreteToContinuous(contA, contB, dt);
```

Comment:
I'm a little confused why we have one constructor that takes a discrete system and makes it continuous, and another that goes the other direction, but no constructor that takes a discrete system and _keeps_ it discrete.

Reply:
The `createDiscrete()` method only stores the passed in discrete system and input matrices. The reason that we need to convert the system matrix into continuous form temporarily is because the passed in process covariance matrix is continuous. Discretizing the process covariance matrix requires a continuous system matrix, which requires us to temporarily convert the system matrix into continuous form.